feat: add find and replace for passport variable via cli prompts

Mike-Heneghan opened the pull request:
## What

- Reuse existing logic to build a script which takes in the current version of the passport variable and the new value it should be replaced with

## Why

- First use case would be migrating from using `outbuildings` to `outbuilding` 
- Opted to make this more abstract to cover potential similar use cases.

## Testing

Local:
- [x] Simple flow testing 
- [ ] More complex flows (sequencing of portals with main services?)

Staging:
- [ ] Simple flow testing 
- [ ] More complex flows (sequencing of portals with main services?)

## Screenshots

<img width="780" alt="Screenshot 2024-02-19 at 14 24 58" src="https://github.com/theopensystemslab/planx-data-migrations/assets/36415632/53bb75d9-2ff7-4c07-ba91-37bb557d973e">

`outbuildings → outbuilding`

<img width="780" alt="Screenshot 2024-02-19 at 14 31 41" src="https://github.com/theopensystemslab/planx-data-migrations/assets/36415632/b1f66749-df97-4346-a748-f65c61edc476">

`outbuilding.shed → shed`

<img width="780" alt="Screenshot 2024-02-19 at 14 32 14" src="https://github.com/theopensystemslab/planx-data-migrations/assets/36415632/66bfc818-c53c-42c6-b92d-90d19b81fe71">

Mike-Heneghan committed:
feat: add find and replace for passport variable via cli prompts

Mike-Heneghan commented on findAndReplacePassportValue/helpers.js:
```diff
@@ -0,0 +1,23 @@
+      const currentFn = nodeData["data"][`${passportKey}`]
+      const currentValueSegments = currentFn.split(".")
+      const matchIndexPosition = currentValueSegments.indexOf(currentPassportVariable)
+      if (matchIndexPosition !==-1) {
```
Due to this line the script will only ever replace the first instance of the exact match string. 

I don't think there'd ever be cases of something like `extend.outbuildings.shed.outbuildings` but if that's not the case I can review the logic here.

jessicamcinchak replied:
If we look at real ProjectTypes, we can see that there are edge cases where the same word can appear twice - eg https://github.com/theopensystemslab/digital-planning-data-schemas/blob/main/types/enums/ProjectTypes.ts#L98

Is it actually necessary to `split(".")` the fn/val string here and then re-join, can we simply try a `replaceAll()` on the full passport string - which would then also give us more flexibility if we want to use this script ot replace _full_ passport variables (any number characters/segments) rather than only a single segment of them?

Mike-Heneghan replied:
Thanks for sharing the `projectTypes` and checking the assumption I'd made!

I'll have another look at this functionality. I had been using `.replace` the reason I was splitting was so I could do exact matches as I was worried about incorrectly replacing substrings so for something like `alter → change` incorrectly leadings to`alteration.openings → changeation.openings`

I'll have another look at `replaceAll` to see how whole sections can be updated 👍

Mike-Heneghan replied:
Fixed here: 4bd01282d2359b91010b784c76374f7d879f32ff

jessicamcinchak replied:
Oops sorry left this as a new comment rather than on this thread: https://github.com/theopensystemslab/planx-data-migrations/pull/9#discussion_r1494585500

Mike-Heneghan committed:
chore: add missing start script

jessicamcinchak commented on package.json:
```diff
@@ -7,6 +7,7 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "start:fileNames": "node fileNames/index.js",
     "start:titleBoundary": "node titleBoundary/index.js",
+    "start:findAndReplacePassportValue": "node findAndReplacePassportValue/index.js",
```
nit: can this subdirectory and therefore its commands just be `passportValues` for consistency/simplicity? `findAndReplace` is implicit across this whole repo I think?

Mike-Heneghan replied:
That's a great idea, I was thinking the name was way too clunky when I was writing it.

Mike-Heneghan replied:
Fixed here: 7fcb89b466ccf012bc60881a855e341a42f59c67

jessicamcinchak commented on findAndReplacePassportValue/prompts.js:
```diff
@@ -0,0 +1,22 @@
+    default: "outbuilding",
+    type: "string",
+    required: true,
+  },
```
:+1: similar to comment above - these are currently phrased as if I can pass a _full_ passport variable to replace (for example `alter.opening` &rarr; `alteration.openings`) 

But the logic in the replace function actually only works on _segments_ of variables. Would be great to be consistent and preference flexibility of replacing any number of characters independent of segments please!

jessicamcinchak reviewed (commented):
Looking good - a few small comments ! 

Happy to test locally again once those are resolved, then let's discuss which flows this should be run against on staging so that August & George can also test :+1:

Mike-Heneghan committed:
fix: update replace to match all occurrences and update values rather than segments

Mike-Heneghan committed:
chore: rename subdirectory

jessicamcinchak commented on passportValue/helpers.js:
```diff
@@ -0,0 +1,22 @@
+
+function replaceAllOccurrences(fullPassportValue, currentPassportVariable, newPassportVariable) {
+  const regex = new RegExp('\\b' + currentPassportVariable + '\\b', 'g');
+  return fullPassportValue.replace(regex, newPassportVariable);
```
Sorry - still not personally convinced this regex that protects against partial matches is better approach than plain string `replaceAll()` and I think is going to be ultimately limiting in use of this against "real" content examples!

In your example `alter` &rarr; `change` (but _not_ `alteration` &rarr; `changeation`) that this regex protects against, I think you're overlooking the option to simply pass `alter.` as the find variable then `replaceAll()` as expected !! 

In recent file naming, there were other examples of partial matches like `councilTaxBill` (find = `councilTax`) &rarr; `councilBill` (replace = `council`) that wouldn't be supported by this regex.

Mike-Heneghan replied:
I'm maybe being too risk averse with this approach. As far as I understand the regex would approach would still allow all renaming but might require more runs as matches would need to be made explicitly `councilTaxBill → councilBill` vs the freedom / risk of partial match renames. 

Passing in variables with `.` as in `alter.` is definitely something I hadn't considered and does overcome the issue I'd outlined although I guess it does require protecting against edge cases at the time of running the code and a deeper familiarity with the project types versus having that edge case protection in the code? 

Although, I'm happy to defer to your judgement though as you're much more familiar with this process and the data.

Mike-Heneghan replied:
Updated here: f073806941afb401f0575a2a4c8eb8afa649d7df

jessicamcinchak replied:
Thanks ! My theory on these is: always start quick & flexible based on known content examples, and introduce strictness as future use cases necessitate. Because since these scripts are only manually run in very controlled scenarios right now (eg not exposed to editors), we don't need to get hung up anticipating uses beyond the task at hand!

Mike-Heneghan committed:
refactor: update from regex replace to replaceAll

- As per: https://github.com/theopensystemslab/planx-data-migrations/pull/9#discussion_r1494585500
- Allows more flexibility

jessicamcinchak reviewed (approved):
Thanks again for picking this one up! Happy for code to merge to `main` whenever you're ready. 

As it's already a bit late in the afternoon & George isn't back until tomorrow, I'd advise running the outbuilding find&replace against staging tomorrow morning for final review from August & George (will get overwritten by nightly sync). 

We'll want to run it against all flows that ask questions about project type (in addition to any parent services that nest those flows as external portals) - which I expect will be a quite long list I haven't started to compile yet. I can take over testing from here if helpful?

Once content team reviews staging & confirms automations are working as expected, I'll make the schema changes and run against production after hours later this week.

Mike-Heneghan commented:
> Thanks again for picking this one up! Happy for code to merge to `main` whenever you're ready.
> 
> As it's already a bit late in the afternoon & George isn't back until tomorrow, I'd advise running the outbuilding find&replace against staging tomorrow morning for final review from August & George (will get overwritten by nightly sync).
> 
> We'll want to run it against all flows that ask questions about project type (in addition to any parent services that nest those flows as external portals) - which I expect will be a quite long list I haven't started to compile yet. I can take over testing from here if helpful?
> 
> Once content team reviews staging & confirms automations are working as expected, I'll make the schema changes and run against production after hours later this week.

Thanks for the review @jessicamcinchak and as advised I'll try running outbuilding find&replace against staging tomorrow morning 👍 

Yeah I'm still not sure how best to go about forming that exhaustive list of flows to update so if you don't mind picking it up from there that would be great.

jessicamcinchak commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Find & replace "outbuildings" to "outbuilding" across project selection questions to fix flow automation and schema consistency](https://trello.com/c/K7OUyI8V/2778-find-replace-outbuildings-to-outbuilding-across-project-selection-questions-to-fix-flow-automation-and-schema-consistency)
